distributed/comm/asyncio_tcp.py

```diff
@@ -0,0 +1,969 @@
+            frame[
+                -self._frame_nbytes_needed : (n_read - self._frame_nbytes_needed)
+                or None
+            ] = self._default_buffer[self._default_start : self._default_start + n_read]
+            self._default_start += n_read
+            self._frame_nbytes_needed -= n_read
+
+    def _reset_default_buffer(self):
+        """Reset the default buffer for the next read event"""
+        start = self._default_start
+        end = self._default_end
+
+        if start < end and start != 0:
+            # Still some unparsed data, copy it to the front of the buffer
+            self._default_buffer[: end - start] = self._default_buffer[start:end]
+            self._default_start = 0
+            self._default_end = end - start
+        elif start == end:
+            # All data is parsed, just reset the indices
+            self._default_start = 0
+            self._default_end = 0
+
+    def _message_completed(self):
+        """Push a completed message to the queue and reset per-message state"""
+        self._queue.put_nowait(self._frames)
+        self._nframes = None
+        self._frames = None
+        self._frame_lengths = None
+        self._frame_nbytes_remaining = 0
+
+    def connection_lost(self, exc=None):
+        self._transport = None
+        self._closed_waiter.set_result(None)
+
+        # Unblock read, if any
+        self._queue.put_nowait(_COMM_CLOSED)
+
+        # Unblock write, if any
+        if self._paused:
+            waiter = self._drain_waiter
+            if waiter is not None:
+                self._drain_waiter = None
+                if not waiter.done():
+                    waiter.set_exception(CommClosedError("Connection closed"))
+
+    def pause_writing(self):
+        self._paused = True
+
+    def resume_writing(self):
+        self._paused = False
+
+        waiter = self._drain_waiter
+        if waiter is not None:
+            self._drain_waiter = None
+            if not waiter.done():
+                waiter.set_result(None)
+
+    async def read(self) -> list[bytes]:
+        """Read a single message from the comm."""
+        # Even if comm is closed, we still yield all received data before
+        # erroring
+        if self._queue is not None:
+            out = await self._queue.get()
+            if out is not _COMM_CLOSED:
+                return out
+            self._queue = None
+        raise CommClosedError("Connection closed")
+
+    async def write(self, frames: list[bytes]) -> int:
+        """Write a message to the comm."""
+        if self.is_closed:
+            raise CommClosedError("Connection closed")
+        elif self._paused:
+            # Wait until there's room in the write buffer
+            drain_waiter = self._drain_waiter = self._loop.create_future()
+            await drain_waiter
+
+        # Ensure all memoryviews are in single-byte format
+        frames = [f.cast("B") if isinstance(f, memoryview) else f for f in frames]
```

Comment:
Are there any situations where a frame wouldn't be a `memoryview`? Should we just coercing everything to `memoryview`s upfront?

Reply:
> Are there any situations where a frame wouldn't be a memoryview?

Yes, `to_frames` doesn't always return memoryviews.

> Should we just coercing everything to memoryviews upfront?

I don't think so, some backing loops have different paths for `bytes` and non-`bytes` objects. The `_ZeroCopyWriter` implemented below will coerce everything to a memoryview itself, but we shouldn't force that on every implementation.